print("============================================================")
print("Description:")
print("TREK analysis: assigning gRNA labels to cells")
print("Compulsory Arguments:")
print("--workDir= full working directory where results will be generated")
print("--ExpFile= full path to pre-processed expression matrix")
print("--BarcodeFile= full path to barcode alignment results generated by barcdode aligning script")
print("--AnnoFile= full path to barcode annotation file")
print("--jobName= name of the CellRanger run")
print("Optional Arguments:")
print("--DominanceRatio= ratio threshold for a gRNA to be considered as dominant n default is 5")
print("--UMIthreshold= minimum UMI count for a gRNA to be considered")
print("=============================================================")
print("\n")
#manual

args <- commandArgs(trailingOnly = T)
###########################################################################
### full working directory where results will be generated
workDir <- args[grep("--workDir=", args)] 
workDir <- substr(workDir, 11, nchar(workDir))
print("Working directory:")
print(workDir)
###########################################################################
### full path to pre-processed expression matrix
ExpFile <- args[grep("--ExpFile=", args)] 
ExpFile <- substr(ExpFile, 11, nchar(ExpFile))
print("ExpFile:")
print(ExpFile)
###########################################################################
### full path to barcode alignment results generated by barcdode aligning script
CROPseqBCFile <- args[grep("--CROPseqBCFile=", args)] 
CROPseqBCFile <- substr(CROPseqBCFile, 17, nchar(CROPseqBCFile))
print("CROPseqBCFile:")
print(CROPseqBCFile)
###########################################################################
###/full/path/to/barcode/alignment/results/generated/by/barcdode/aligning/script
CellRangerBCFile <- args[grep("--CellRangerBCFile=", args)] 
CellRangerBCFile <- substr(CellRangerBCFile, 20, nchar(CellRangerBCFile))
print("CellRangerBCFile:")
print(CellRangerBCFile)
###########################################################################
### full path to barcode alignment results generated by barcdode aligning script
AnnoFile <- args[grep("--AnnoFile=", args)] 
AnnoFile <- substr(AnnoFile, 12, nchar(AnnoFile))
print("AnnoFile:")
print(AnnoFile)
###########################################################################
### name of the CROP-seq run
jobName <- args[grep("--jobName=", args)] 
jobName <- substr(jobName, 11, nchar(jobName))
print("Job name:")
print(jobName)
##########################################################################
### ratio threshold for a gRNA to be considered as dominant default is 5
DominanceRatio <- args[grep("--DominanceRatio=", args)] 
DominanceRatio <- as.numeric(substr(DominanceRatio, 18, nchar(DominanceRatio)))
print("Dominance ratio:")
print(DominanceRatio)
##########################################################################
### minimum UMI count for a gRNA to be considered
UMIthreshold <- args[grep("--UMIthreshold=", args)] 
UMIthreshold <- as.numeric(substr(UMIthreshold, 16, nchar(UMIthreshold)))
print("UMI threshold of gRNA calling:")
print(UMIthreshold)
##########################################################################
### maximum mismatch allowed for cell barcode
CellBCorrection <- args[grep("--CellBCorrection=", args)] 
CellBCorrection <- as.numeric(substr(CellBCorrection, 19, nchar(CellBCorrection)))
print("Maximum mismatch allowed for cell barcode, at most 3:")
print(CellBCorrection)
#arguments

##############################################################################################
### Function: Ratio_gRNA_assign
### Assign gRNA to cells using Ratio method.
### Args:
###       gRNA_exprs_mat: gRNA expression matrix
###       ratio: minimum ratio between dominant gRNA and minor gRNAs, default=5
###       UMI_thres: minimum UMI count for a gRNA to be considered, defualt=2
Ratio_gRNA_assign <- function(gRNA_exprs_mat,ratio=5,UMI_thres=2){
  assignment <- apply(gRNA_exprs_mat, 2, function(v){
    v <- sort(v,decreasing = T)
    # if no gRNA got >= [UMI_thres] read, consider gRNA_No.=0 #
    if (v[1]<UMI_thres) return(c())
    else {
      # if there is a dominant gRNA (at least [ratio] times more reads than second gRNA) #
      ans = c(names(v)[1])
      if (v[1]>(ratio*v[2])) return(ans)
      # multiple gRNAs #
      else {
        i=2
        while ((v[1]<=(v[i]*ratio))&(v[i]>=UMI_thres)) {
          ans=c(ans,names(v)[i])
          i=i+1
        }
        return(ans)
      }
    }
  })
  return(assignment)
}
##############################################################################################
### Function: Plot_gRNA_stackHist_UMI
### Make stack histograms of gRNA No. distribution by UMI per cell.
### Args:
###       exprs: single-cell expression matrix
###       gRNA_assigned: gRNA labels associated with the cells in exprs
Plot_gRNA_stackHist_UMI <- function(exprs,gRNA_assigned){
  gRNA_No._assigned<-sapply(gRNA_assigned, length)
  guide_No._per_cell <- gRNA_No._assigned
  UMI_per_cell <- apply(exprs, 2, sum)
  UMI_tiers <- round(quantile(UMI_per_cell,(0:6/6)),-3)
  assign_UMI<- data.frame(
    UMI_tier_1 = c(sum(guide_No._per_cell==0&UMI_per_cell>UMI_tiers[1]&UMI_per_cell<=UMI_tiers[2])/sum(UMI_per_cell>UMI_tiers[1]&UMI_per_cell<=UMI_tiers[2]),
                   sum(guide_No._per_cell==1&UMI_per_cell>UMI_tiers[1]&UMI_per_cell<=UMI_tiers[2])/sum(UMI_per_cell>UMI_tiers[1]&UMI_per_cell<=UMI_tiers[2]),
                   sum(guide_No._per_cell>1&UMI_per_cell>UMI_tiers[1]&UMI_per_cell<=UMI_tiers[2])/sum(UMI_per_cell>UMI_tiers[1]&UMI_per_cell<=UMI_tiers[2])),
    UMI_tier_2 = c(sum(guide_No._per_cell==0&UMI_per_cell>UMI_tiers[2]&UMI_per_cell<=UMI_tiers[3])/sum(UMI_per_cell>UMI_tiers[2]&UMI_per_cell<=UMI_tiers[3]),
                   sum(guide_No._per_cell==1&UMI_per_cell>UMI_tiers[2]&UMI_per_cell<=UMI_tiers[3])/sum(UMI_per_cell>UMI_tiers[2]&UMI_per_cell<=UMI_tiers[3]),
                   sum(guide_No._per_cell>1&UMI_per_cell>UMI_tiers[2]&UMI_per_cell<=UMI_tiers[3])/sum(UMI_per_cell>UMI_tiers[2]&UMI_per_cell<=UMI_tiers[3])),
    UMI_tier_3 = c(sum(guide_No._per_cell==0&UMI_per_cell>UMI_tiers[3]&UMI_per_cell<=UMI_tiers[4])/sum(UMI_per_cell>UMI_tiers[3]&UMI_per_cell<=UMI_tiers[4]),
                   sum(guide_No._per_cell==1&UMI_per_cell>UMI_tiers[3]&UMI_per_cell<=UMI_tiers[4])/sum(UMI_per_cell>UMI_tiers[3]&UMI_per_cell<=UMI_tiers[4]),
                   sum(guide_No._per_cell>1&UMI_per_cell>UMI_tiers[3]&UMI_per_cell<=UMI_tiers[4])/sum(UMI_per_cell>UMI_tiers[3]&UMI_per_cell<=UMI_tiers[4])),
    UMI_tier_4 = c(sum(guide_No._per_cell==0&UMI_per_cell>UMI_tiers[4]&UMI_per_cell<=UMI_tiers[5])/sum(UMI_per_cell>UMI_tiers[4]&UMI_per_cell<=UMI_tiers[5]),
                   sum(guide_No._per_cell==1&UMI_per_cell>UMI_tiers[4]&UMI_per_cell<=UMI_tiers[5])/sum(UMI_per_cell>UMI_tiers[4]&UMI_per_cell<=UMI_tiers[5]),
                   sum(guide_No._per_cell>1&UMI_per_cell>UMI_tiers[4]&UMI_per_cell<=UMI_tiers[5])/sum(UMI_per_cell>UMI_tiers[4]&UMI_per_cell<=UMI_tiers[5])),
    UMI_tier_5 = c(sum(guide_No._per_cell==0&UMI_per_cell>UMI_tiers[5]&UMI_per_cell<=UMI_tiers[6])/sum(UMI_per_cell>UMI_tiers[5]&UMI_per_cell<=UMI_tiers[6]),
                   sum(guide_No._per_cell==1&UMI_per_cell>UMI_tiers[5]&UMI_per_cell<=UMI_tiers[6])/sum(UMI_per_cell>UMI_tiers[5]&UMI_per_cell<=UMI_tiers[6]),
                   sum(guide_No._per_cell>1&UMI_per_cell>UMI_tiers[5]&UMI_per_cell<=UMI_tiers[6])/sum(UMI_per_cell>UMI_tiers[5]&UMI_per_cell<=UMI_tiers[6])),
    UMI_tier_6 = c(sum(guide_No._per_cell==0&UMI_per_cell>UMI_tiers[6]&UMI_per_cell<=UMI_tiers[7])/sum(UMI_per_cell>UMI_tiers[6]&UMI_per_cell<=UMI_tiers[7]),
                   sum(guide_No._per_cell==1&UMI_per_cell>UMI_tiers[6]&UMI_per_cell<=UMI_tiers[7])/sum(UMI_per_cell>UMI_tiers[6]&UMI_per_cell<=UMI_tiers[7]),
                   sum(guide_No._per_cell>1&UMI_per_cell>UMI_tiers[6]&UMI_per_cell<=UMI_tiers[7])/sum(UMI_per_cell>UMI_tiers[6]&UMI_per_cell<=UMI_tiers[7]))
  )
  pdf(file = paste0("gRNA_StackHist_byUMI_",jobName,".pdf"),width = 14,height = 14)
  par(lwd=2,xpd=T,mar=c(5,5,2,10))
  barplot(as.matrix(assign_UMI)[c(2,1,3),],col = c("cyan3","gray","firebrick3"),lwd=2,ylab="Percentage of cells",cex.lab=1.5,xaxt='n',xlab="UMI_per_cell")
  axis(1,at=(1:7)/0.833-1.09,lwd=2,labels = UMI_tiers)
  legend("topright",fill = c("firebrick3","gray","cyan3"),title="Cells",inset = c(-0.17,0),
         legend = c("multiple-gRNA cell","zero-gRNA cells","unique-gRNA cells"))
  dev.off()
}
##############################################################################################
### Function: Plot_gRNA_stackHist_Gene
### Make stack histograms of gRNA No. distribution by Gene per cell.
### Args:
###       exprs: single-cell expression matrix
###       gRNA_assigned: gRNA labels associated with the cells in exprs
Plot_gRNA_stackHist_Gene <- function(exprs,gRNA_assigned){
  gRNA_No._assigned<-sapply(gRNA_assigned, length)
  guide_No._per_cell <- gRNA_No._assigned
  gene_per_cell <- apply(exprs, 2, function(v){sum(v>1)})
  gene_tiers <- round(quantile(gene_per_cell,(0:6/6)),-2)
  assign_gene<- data.frame(
    gene_tier_1 = c(sum(guide_No._per_cell==0&gene_per_cell>gene_tiers[1]&gene_per_cell<=gene_tiers[2])/sum(gene_per_cell>gene_tiers[1]&gene_per_cell<=gene_tiers[2]),
                    sum(guide_No._per_cell==1&gene_per_cell>gene_tiers[1]&gene_per_cell<=gene_tiers[2])/sum(gene_per_cell>gene_tiers[1]&gene_per_cell<=gene_tiers[2]),
                    sum(guide_No._per_cell>1&gene_per_cell>gene_tiers[1]&gene_per_cell<=gene_tiers[2])/sum(gene_per_cell>gene_tiers[1]&gene_per_cell<=gene_tiers[2])),
    gene_tier_2 = c(sum(guide_No._per_cell==0&gene_per_cell>gene_tiers[2]&gene_per_cell<=gene_tiers[3])/sum(gene_per_cell>gene_tiers[2]&gene_per_cell<=gene_tiers[3]),
                    sum(guide_No._per_cell==1&gene_per_cell>gene_tiers[2]&gene_per_cell<=gene_tiers[3])/sum(gene_per_cell>gene_tiers[2]&gene_per_cell<=gene_tiers[3]),
                    sum(guide_No._per_cell>1&gene_per_cell>gene_tiers[2]&gene_per_cell<=gene_tiers[3])/sum(gene_per_cell>gene_tiers[2]&gene_per_cell<=gene_tiers[3])),
    gene_tier_3 = c(sum(guide_No._per_cell==0&gene_per_cell>gene_tiers[3]&gene_per_cell<=gene_tiers[4])/sum(gene_per_cell>gene_tiers[3]&gene_per_cell<=gene_tiers[4]),
                    sum(guide_No._per_cell==1&gene_per_cell>gene_tiers[3]&gene_per_cell<=gene_tiers[4])/sum(gene_per_cell>gene_tiers[3]&gene_per_cell<=gene_tiers[4]),
                    sum(guide_No._per_cell>1&gene_per_cell>gene_tiers[3]&gene_per_cell<=gene_tiers[4])/sum(gene_per_cell>gene_tiers[3]&gene_per_cell<=gene_tiers[4])),
    gene_tier_4 = c(sum(guide_No._per_cell==0&gene_per_cell>gene_tiers[4]&gene_per_cell<=gene_tiers[5])/sum(gene_per_cell>gene_tiers[4]&gene_per_cell<=gene_tiers[5]),
                    sum(guide_No._per_cell==1&gene_per_cell>gene_tiers[4]&gene_per_cell<=gene_tiers[5])/sum(gene_per_cell>gene_tiers[4]&gene_per_cell<=gene_tiers[5]),
                    sum(guide_No._per_cell>1&gene_per_cell>gene_tiers[4]&gene_per_cell<=gene_tiers[5])/sum(gene_per_cell>gene_tiers[4]&gene_per_cell<=gene_tiers[5])),
    gene_tier_5 = c(sum(guide_No._per_cell==0&gene_per_cell>gene_tiers[5]&gene_per_cell<=gene_tiers[6])/sum(gene_per_cell>gene_tiers[5]&gene_per_cell<=gene_tiers[6]),
                    sum(guide_No._per_cell==1&gene_per_cell>gene_tiers[5]&gene_per_cell<=gene_tiers[6])/sum(gene_per_cell>gene_tiers[5]&gene_per_cell<=gene_tiers[6]),
                    sum(guide_No._per_cell>1&gene_per_cell>gene_tiers[5]&gene_per_cell<=gene_tiers[6])/sum(gene_per_cell>gene_tiers[5]&gene_per_cell<=gene_tiers[6])),
    gene_tier_6 = c(sum(guide_No._per_cell==0&gene_per_cell>gene_tiers[6]&gene_per_cell<=gene_tiers[7])/sum(gene_per_cell>gene_tiers[6]&gene_per_cell<=gene_tiers[7]),
                    sum(guide_No._per_cell==1&gene_per_cell>gene_tiers[6]&gene_per_cell<=gene_tiers[7])/sum(gene_per_cell>gene_tiers[6]&gene_per_cell<=gene_tiers[7]),
                    sum(guide_No._per_cell>1&gene_per_cell>gene_tiers[6]&gene_per_cell<=gene_tiers[7])/sum(gene_per_cell>gene_tiers[6]&gene_per_cell<=gene_tiers[7]))
  )
  pdf(file = paste0("gRNA_StackHist_byGene_",jobName,".pdf"),width = 14,height = 14)
  par(lwd=2,xpd=T,mar=c(5,5,2,10))
  barplot(as.matrix(assign_gene)[c(2,1,3),],col = c("cyan3","gray","firebrick3"),lwd=2,ylab="Percentage of cells",cex.lab=1.5,xaxt='n',xlab="gene_per_cell")
  axis(1,at=(1:7)/0.833-1.09,lwd=2,labels = gene_tiers)
  legend("topright",fill = c("firebrick3","gray","cyan3"),title="Cells",inset = c(-0.17,0),
         legend = c("multiple-gRNA cell","zero-gRNA cells","unique-gRNA cells"))
  dev.off()
}
##############################################################################################
Plot_gRNAp_per_Cell <- function(exprs_gRNA,UMIthreshold,DominanceRatio,source){
  # Without consider dominance
  gRNA_No. <- apply(exprs_gRNA, 2, function(v){length(which(v>=UMIthreshold))})
  # Plot histogram.
  pdf(paste0("NumberOfGuideRNAPerCell_WithoutDominance_",source,".pdf"),width=7,height=7)
  par(lwd=2)
  hist(gRNA_No.,probability = T,right = F,ylim = c(0,1),lwd=2,font=2,cex.lab=1.5,xaxt='n',
       main="Number of gRNA per cell - without considering dominance",xlab = "Number of gRNA",
       breaks = 0:(max(gRNA_No.)+1))
  axis(1,at=0:(max(gRNA_No.)+1),lwd=2,font=2,labels = F)
  axis(1,at=0:max(gRNA_No.)+0.5,lwd=2,font=2,labels = 0:max(gRNA_No.),tick = 0)
  dev.off()
  # Considering dominance
  gRNA_assigned <- Ratio_gRNA_assign(exprs_gRNA,ratio = DominanceRatio,UMI_thres = UMIthreshold)
  gRNA_No._assigned<-sapply(gRNA_assigned, length)
  print(paste0("No. of unique gRNA cells: ",sum(gRNA_No._assigned==1)))
  print(paste0("No. of unidentifiable cells: ",sum(gRNA_No._assigned==0)))
  print(paste0("No. of multi-gRNA cells: ",sum(gRNA_No._assigned>1)))
  # Plot histogram.
  pdf(paste0("NumberOfGuideRNAPerCell_WithDominance_",source,".pdf"),width=7,height=7)
  par(lwd=2)
  hist(gRNA_No._assigned, probability = T,right = F,ylim = c(0,1),lwd=2,font=2,cex.lab=1.5,breaks = 0:(max(gRNA_No._assigned)+1),
       main="Number of gRNA per cell - considering dominance",xlab = "Number of gRNA",xaxt='n')
  axis(1,at=0:(max(gRNA_No._assigned)+1),lwd=2,font=2,labels = F)
  axis(1,at=0:max(gRNA_No._assigned)+0.5,lwd=2,font=2,labels = 0:max(gRNA_No._assigned),tick = 0)
  dev.off()
}
##############################################################################################
gRNA_Consolidate <- function(exprs_gRNA_CROPseq,exprs_gRNA_CellRanger){
  gRNA_assign_CROPseq <- Ratio_gRNA_assign(exprs_gRNA_CROPseq,DominanceRatio,UMIthreshold)
  gRNA_assign_CellRanger <- Ratio_gRNA_assign(exprs_gRNA_CellRanger,DominanceRatio,UMIthreshold)
  gRNA_assign <- sapply(names(gRNA_assign_CROPseq), function(c){
    if(is.null(gRNA_assign_CROPseq[[c]]) & is.null(exprs_gRNA_CellRanger[[c]])) return(NULL)
    if(is.null(gRNA_assign_CROPseq[[c]]) & !is.null(exprs_gRNA_CellRanger[[c]])) return(exprs_gRNA_CellRanger[[c]])
    if(!is.null(gRNA_assign_CROPseq[[c]]) & is.null(exprs_gRNA_CellRanger[[c]])) return(gRNA_assign_CROPseq[[c]])
    if(length(gRNA_assign_CROPseq[[c]])==1 & length(exprs_gRNA_CellRanger[[c]])==1) {
      if(gRNA_assign_CROPseq[[c]] == exprs_gRNA_CellRanger[[c]]) return(gRNA_assign_CROPseq[[c]])
      else return(c(gRNA_assign_CROPseq[[c]],exprs_gRNA_CellRanger[[c]]))
    }
    if(length(gRNA_assign_CROPseq[[c]])>1 | length(exprs_gRNA_CellRanger[[c]])>1)
      return(unique(c(gRNA_assign_CROPseq[[c]],exprs_gRNA_CellRanger[[c]])))
  })
}



setwd(workDir)


##################################### Count CROPseq gRNA counts #######################################
load(ExpFile)
### Correcting chimeric barcode enrichemment ################################################
cells <- colnames(exprs)
# Load chimeric-corrected library.
processed_barcode <- as.matrix(read.table(CROPseqBCFile,header = T))
# Get rid of unmapped barcode.
processed_barcode <- processed_barcode[-grep("unprocessed",processed_barcode[,2]),]
# Get rid of undetected cells.
unique_cells <- unique(processed_barcode[,1])
match <- sapply(unique_cells, function(s){
  if (s%in%cells) return(s)
  ham.dist <- sapply(cells, function(c){e1071::hamming.distance(unlist(strsplit(s,"")),unlist(strsplit(c,"")))})
  if (CellBCorrection>0) if (sum(ham.dist==1)==1) return(cells[ham.dist==1])
  if (CellBCorrection>1) if ((sum(ham.dist==1)==0)&(sum(ham.dist==2)==1)) return(cells[ham.dist==2])
  if (CellBCorrection>2) if ((sum(ham.dist==1)==0)&(sum(ham.dist==2)==0)&(sum(ham.dist==3)==1)) return(cells[ham.dist==3])
  else return(NULL)
})
unique_cells <- unique_cells[!sapply(match,is.null)]
match <- as.character(match[!sapply(match,is.null)])
processed_barcode <- processed_barcode[processed_barcode[,1]%in%unique_cells,]
names(match) <- unique_cells

# Load gRNA annotation file.
barcode_annotation <- as.matrix(read.table(AnnoFile))
# Map.
barcode_gRNA <- sapply(processed_barcode[,2], function(b){
  anno <- barcode_annotation[grep(b,toupper(barcode_annotation[,2])),1]
  if (length(anno)>1) return(anno[1])
  return(anno)
})
processed_barcode <- cbind(processed_barcode,paste0(barcode_gRNA,""))

# Adding to gRNA_exprs.
exprs_gRNA_CROPseq <- matrix(0,nrow = nrow(barcode_annotation),ncol = length(cells))
rownames(exprs_gRNA_CROPseq) <- paste0(barcode_annotation[,1],"")
colnames(exprs_gRNA_CROPseq) <- cells
for (i in 1:nrow(processed_barcode)) {
  #print(processed_barcode[i,5])
  exprs_gRNA_CROPseq[processed_barcode[i,5],match[processed_barcode[i,1]]] <- as.numeric(processed_barcode[i,4])
}
Plot_gRNAp_per_Cell(exprs_gRNA_CROPseq,UMIthreshold,DominanceRatio,"CROPseq")
#############################################################################################

CellRanger_Call <- as.matrix(read.csv(CellRangerBCFile))
barcode_annotation <- as.matrix(read.table(AnnoFile))
load(ExpFile)
exprs_gRNA_CellRanger <- matrix(0,nrow(barcode_annotation),ncol=ncol(exprs))
rownames(exprs_gRNA_CellRanger) <- barcode_annotation[,1]
colnames(exprs_gRNA_CellRanger) <- colnames(exprs)
for (i in 1:nrow(CellRanger_Call)) {
  gRNAs <- as.numeric(gsub("TREK_SNU16_","",unlist(strsplit(as.character(CellRanger_Call[i,3]),split = "\\|"))))
  umis <- as.numeric(unlist(strsplit(as.character(CellRanger_Call[i,4]),split = "\\|")))
  for (j in 1:CellRanger_Call[i,2]) {
    if(as.character(CellRanger_Call[i,1])%in%colnames(exprs_gRNA_CellRanger))
      exprs_gRNA_CellRanger[gRNAs[j],as.character(CellRanger_Call[i,1])] <- umis[j]
  }
}
Plot_gRNAp_per_Cell(exprs_gRNA_CellRanger,UMIthreshold,DominanceRatio,"CellRanger")
##############################################################################################
gRNA_assigned <- gRNA_Consolidate(exprs_gRNA_CROPseq,exprs_gRNA_CellRanger)



### By how many 1-gRNA&dCas9 cells each gRNA is covered? #####################################
gRNA_coverage <- rep(0,times=nrow(barcode_annotation))
names(gRNA_coverage) <- barcode_annotation[,1]
for (c in gRNA_assigned[gRNA_No._assigned==1]) gRNA_coverage[c]=gRNA_coverage[c] +1
pdf(paste0("NumberOfUniqueGuideCellsPerGuide_",jobName,".pdf"),width = 7,height = 7)
par(lwd=2)
# Plot histogram.
hist(gRNA_coverage, 
     probability = T,right = F,breaks = 0:max(gRNA_coverage+1),ylim = c(0,0.3),lwd=2,font=2,cex.lab=1.5,xaxt='n',
     main="Number of unique-gRNA cells per gRNA",xlab = "Number of unique-gRNA cells")
axis(1,at=0:(max(gRNA_coverage)+1),lwd=2,font=2,labels = F)
axis(1,at=0:max(gRNA_coverage)+0.5,lwd=2,font=2,labels = 0:max(gRNA_coverage),tick = 0)
dev.off()

### Save gRNA assignment #####################################################################

save(gRNA_assigned,exprs_gRNA_CROPseq,exprs_gRNA_CellRanger,file = paste0(jobName,".rda"))
##############################################################################################

Plot_gRNA_stackHist_Gene(exprs,gRNA_assigned)
Plot_gRNA_stackHist_UMI(exprs,gRNA_assigned)

